core/trino-main/src/main/java/io/trino/operator/DuplicateRowFinder.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.spi.Page;
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators;
+import io.trino.type.BlockTypeOperators.BlockPositionIsDistinctFrom;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.spi.StandardErrorCode.MERGE_TARGET_ROW_MULTIPLE_MATCHES;
+import static io.trino.spi.connector.MergeProcessorUtilities.DEFAULT_CASE_OPERATION_NUMBER;
+import static io.trino.spi.connector.MergeProcessorUtilities.INSERT_OPERATION_NUMBER;
+import static io.trino.spi.type.IntegerType.INTEGER;
+
+/**
+ * This class throws an exception if two adjacent rows have the same identity.
+ * It compares both the rowId column as well as the writeRedistributionColumns,
```

Comment:
`(rowId, writeRedistributionColumns)` semantics are tricky. One could think that `rowId` should suffice, but it's not unique (despite it's name!)

furthermore, i would love to see `rowId` as being opaque and not being interpreted by the engine.
I would love to assume rowId can be of any type (can be number, a bunch of numbers, or a long text), and that it may or may not implement equality operator.

the point class is -- if i understand -- to detect ambiguous UPDATE matches and throw MERGE_TARGET_ROW_MULTIPLE_MATCHES. For this, we just need a known unique column on a target relation. Use `AssignUniqueId` for that.

Reply:
The rowId column **is** opaque from the point of view of the engine, at least in `DuplicateRowFinder`.  The only information it gets about the rowId is the `Type` and the channel number, and it uses the type to do the comparison.

`DeleteAndInsertMergeProcessor` has conditional code if the rowId happens to be a `RowBlock`, but that code is because `RowBlock.getPositions()` is returning unexpected results.  That's a bug to be fixed, maybe in my code or maybe elsewhere.